Performance improvements for add_method_to_attributes!

dillonwelch opened the pull request:
Continuation of https://github.com/rails/rails/pull/30948

Add a cache to save the `method.to_s.downcase` call.

This change prevents a string allocation for each time the method is
called, with performance changes within margin of error.

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "benchmark-ips"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

def add_method_to_attributes!(html_options, method)
  if method && method.to_s.downcase != "get".freeze &&
      html_options["rel".freeze] !~ /nofollow/
    html_options["rel".freeze] =
      "#{html_options["rel".freeze]} nofollow".lstrip
  end
  html_options["data-method".freeze] = method
end

def fast_add_method_to_attributes!(html_options, method)
  @_method_downcase_cache ||= {}
  downcased_method = @_method_downcase_cache[method] ||=
    method.to_s.downcase
  if downcased_method && downcased_method != "get" &&
      html_options["rel"] !~ /nofollow/
    if html_options["rel"].present?
      html_options["rel"] = "#{html_options["rel"]} nofollow"
    else
      html_options["rel"] = "nofollow"
    end
  end
  html_options["data-method"] = downcased_method
end

obj = ''.freeze

[
  [{}, 'GET'],
  [{'rel' => 'hi'}, 'GET'],
  [{}, 'DEL'],
  [{'rel' => 'hi'}, 'DEL']
].each do |html_options, method|
  puts " #{html_options} , #{method}".center(60, '=')

  puts " Memory Usage ".center(60, "=")
  puts

  puts "value.add_method_to_attributes!"
  puts allocate_count {
    1000.times { add_method_to_attributes!(html_options, method) }
  }

  puts "value.fast_add_method_to_attributes!"
  puts allocate_count {
    1000.times { fast_add_method_to_attributes!(html_options, method) }
  }

  puts
  puts " Benchmark.ips ".center(60, "=")
  puts

  Benchmark.ips do |x|
    x.report("add_method_to_attributes!") do
      add_method_to_attributes!(html_options, method)
    end
    x.report("fast_add_method_to_attributes!") do
      fast_add_method_to_attributes!(html_options, method)
    end
    x.compare!
  end
end
```

```
========================= {} , GET==========================
======================= Memory Usage =======================

value.add_method_to_attributes!
{:FREE=>-997, :T_STRING=>1052}
value.fast_add_method_to_attributes!
{:FREE=>-3, :T_STRING=>1, :T_HASH=>1}

====================== Benchmark.ips =======================

Warming up --------------------------------------
add_method_to_attributes!
                       106.245k i/100ms
fast_add_method_to_attributes!
                       105.093k i/100ms
Calculating -------------------------------------
add_method_to_attributes!
                          2.517M (±14.8%) i/s -     12.324M in   5.029593s
fast_add_method_to_attributes!
                          2.441M (±15.7%) i/s -     11.876M in   5.002305s

Comparison:
add_method_to_attributes!:  2517402.3 i/s
fast_add_method_to_attributes!:  2441322.0 i/s - same-ish: difference falls within error

==================== {"rel"=>"hi"} , GET====================
======================= Memory Usage =======================

value.add_method_to_attributes!
{:FREE=>-1001, :T_STRING=>1000}
value.fast_add_method_to_attributes!
{:FREE=>-1}

====================== Benchmark.ips =======================

Warming up --------------------------------------
add_method_to_attributes!
                       122.110k i/100ms
fast_add_method_to_attributes!
                       124.142k i/100ms
Calculating -------------------------------------
add_method_to_attributes!
                          2.640M (±14.9%) i/s -     12.944M in   5.041262s
fast_add_method_to_attributes!
                          2.730M (±11.0%) i/s -     13.531M in   5.029957s

Comparison:
fast_add_method_to_attributes!:  2730043.4 i/s
add_method_to_attributes!:  2640236.3 i/s - same-ish: difference falls within error

========================= {} , DEL==========================
======================= Memory Usage =======================

value.add_method_to_attributes!
{:FREE=>-4003, :T_STRING=>2003, :T_MATCH=>999, :T_IMEMO=>1000}
value.fast_add_method_to_attributes!
{:FREE=>-3003, :T_STRING=>1002, :T_MATCH=>1000, :T_IMEMO=>1000}

====================== Benchmark.ips =======================

Warming up --------------------------------------
add_method_to_attributes!
                        38.073k i/100ms
fast_add_method_to_attributes!
                        35.313k i/100ms
Calculating -------------------------------------
add_method_to_attributes!
                        500.309k (±16.6%) i/s -      2.475M in   5.124336s
fast_add_method_to_attributes!
                        454.153k (±20.2%) i/s -      2.189M in   5.067740s

Comparison:
add_method_to_attributes!:   500309.2 i/s
fast_add_method_to_attributes!:   454152.9 i/s - same-ish: difference falls within error

==================== {"rel"=>"hi"} , DEL====================
======================= Memory Usage =======================

value.add_method_to_attributes!
{:FREE=>-4001, :T_STRING=>2001, :T_MATCH=>999, :T_IMEMO=>1000}
value.fast_add_method_to_attributes!
{:FREE=>-3001, :T_STRING=>1000, :T_MATCH=>1000, :T_IMEMO=>1000}

====================== Benchmark.ips =======================

Warming up --------------------------------------
add_method_to_attributes!
                        36.612k i/100ms
fast_add_method_to_attributes!
                        39.979k i/100ms
Calculating -------------------------------------
add_method_to_attributes!
                        461.501k (±32.8%) i/s -      1.977M in   5.066839s
fast_add_method_to_attributes!
                        517.545k (±13.4%) i/s -      2.559M in   5.048322s

Comparison:
fast_add_method_to_attributes!:   517545.4 i/s
add_method_to_attributes!:   461501.0 i/s - same-ish: difference falls within error
```

### Summary

Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together.

### Other Information

If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails!

oniofchaos commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -589,10 +589,16 @@ def link_to_remote_options?(options)
         def add_method_to_attributes!(html_options, method)
-          if method && method.to_s.downcase != "get".freeze && html_options["rel".freeze] !~ /nofollow/
-            html_options["rel".freeze] = "#{html_options["rel".freeze]} nofollow".lstrip
+          @_method_downcase_cache ||= {}
```
Since this is in a private method, hopefully this doesn't have the same security concern that the dasherize change had? https://github.com/rails/rails/blob/master/actionview/lib/action_view/helpers/url_helper.rb#L577

rafaelfranca replied:
It is still a problem. `link_to :foo, method: params[:method]`. I don't think caching calculation using instance variables is a good idea, even more in a case where the micro benchmarks fall within error.

oniofchaos replied:
Do you have a preference on this? Maybe a constant or something of the common use cases? It looked like this was doing `:get` => `'get'` for each link on the page.

matthewd replied:
Does it show up on a larger-than-micro benchmark?

matthewd replied:
Does `/\Aget\z/i.match?(method)` offer any improvement?

oniofchaos replied:
Just double check this and it was actually `delete` that I was thinking of. This code gets triggered when you pass in a `method` argument to `link_to`. In my use case, it's 1000s of links that have `method: :delete`.

Could add a constant for the standard REST verbs here?

oniofchaos replied:
For reference - with `memory_profiler` and separating out the other change, on a page with 12880 delete links this reduced allocated memory by 1.54MB of string allocations at a cost of a 7.8KB increase in retained memory

oniofchaos commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -589,10 +589,16 @@ def link_to_remote_options?(options)
+            if html_options["rel"].present?
+              html_options["rel"] = "#{html_options["rel"]} nofollow"
+            else
+              html_options["rel"] = "nofollow"
```
Thoughts on this change separate from the cache change? It's also a performance improvement for the case where there is no `html_options["rel"]` value

oniofchaos replied:
Just double checked this and it prevents an allocation of `""` for each `link_to` call that does not pass in a `"rel"` option.

oniofchaos replied:
Moved this to https://github.com/rails/rails/pull/30997

oniofchaos commented:
Rewrote this to use a constant cache of REST verbs instead of a unbounded instance variable cache. Saves two string allocations per method call as well as a 1.5x speedup if `:get` is passed in.

```ruby
# frozen_string_literal: true
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "benchmark-ips"
  gem "rails"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

@html_options = {}

def master_version
  if @method && @method.to_s.downcase != "get" && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

def fast_version
  if method_not_get_method?(@method) && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

STRINGIFIED_COMMON_METHODS = {
  get:    'get',
  delete: 'delete',
  patch:  'patch',
  post:   'post',
  put:    'put',
}.freeze

def method_not_get_method?(method)
  return false unless method
  (STRINGIFIED_COMMON_METHODS[method] || method.to_s.downcase) != 'get'
end

puts 'get'
@method = :get

puts "master_version"
puts allocate_count { 1000.times { master_version } }
puts "fast_version"
puts allocate_count { 1000.times { fast_version } }

Benchmark.ips do |x|
  x.report("master_version") { master_version }
  x.report("fast_version")     { fast_version }
  x.compare!
end

puts 'delete'
@method = :delete

puts "master_version"
puts allocate_count { 1000.times { master_version } }
puts "fast_version"
puts allocate_count { 1000.times { fast_version } }

Benchmark.ips do |x|
  x.report("master_version") { master_version }
  x.report("fast_version")     { fast_version }
  x.compare!
end
```

```
get
master_version
{:FREE=>-1819, :T_STRING=>2052}
fast_version
{:FREE=>-1}
Warming up --------------------------------------
      master_version   140.839k i/100ms
        fast_version   175.639k i/100ms
Calculating -------------------------------------
      master_version      2.683M (± 7.1%) i/s -     13.380M in   5.013447s
        fast_version      3.988M (±10.1%) i/s -     19.847M in   5.039580s

Comparison:
        fast_version:  3988340.3 i/s
      master_version:  2683336.2 i/s - 1.49x  slower

delete
master_version
{:FREE=>-5003, :T_STRING=>3003, :T_MATCH=>999, :T_IMEMO=>1000}
fast_version
{:FREE=>-3002, :T_STRING=>1001, :T_MATCH=>1000, :T_IMEMO=>1000}
Warming up --------------------------------------
      master_version    47.221k i/100ms
        fast_version    44.153k i/100ms
Calculating -------------------------------------
      master_version    597.881k (±11.4%) i/s -      2.975M in   5.047200s
        fast_version    686.014k (±11.6%) i/s -      3.400M in   5.036564s

Comparison:
        fast_version:   686014.5 i/s
      master_version:   597881.4 i/s - same-ish: difference falls within error
```

rafaelfranca commented:
Looks good to me. Can you squash your commits in only one?

dillonwelch committed:
Performance improvements for add_method_to_attributes!

Prevents two string allocations per method call for common REST verbs
plus a ~1.5x speedup for :get in particular

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "benchmark-ips"
  gem "rails"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

@html_options = {}

def master_version
  if @method && @method.to_s.downcase != "get" && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

def fast_version
  if method_not_get_method?(@method) && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

STRINGIFIED_COMMON_METHODS = {
  get:    'get',
  delete: 'delete',
  patch:  'patch',
  post:   'post',
  put:    'put',
}.freeze

def method_not_get_method?(method)
  return false unless method
  (STRINGIFIED_COMMON_METHODS[method] || method.to_s.downcase) != 'get'
end

puts 'get'
@method = :get

puts "master_version"
puts allocate_count { 1000.times { master_version } }
puts "fast_version"
puts allocate_count { 1000.times { fast_version } }

Benchmark.ips do |x|
  x.report("master_version") { master_version }
  x.report("fast_version")     { fast_version }
  x.compare!
end

puts 'delete'
@method = :delete

puts "master_version"
puts allocate_count { 1000.times { master_version } }
puts "fast_version"
puts allocate_count { 1000.times { fast_version } }

Benchmark.ips do |x|
  x.report("master_version") { master_version }
  x.report("fast_version")     { fast_version }
  x.compare!
end
```

```
get
master_version
{:FREE=>-1819, :T_STRING=>2052}
fast_version
{:FREE=>-1}
Warming up --------------------------------------
      master_version   140.839k i/100ms
        fast_version   175.639k i/100ms
Calculating -------------------------------------
      master_version      2.683M (± 7.1%) i/s -     13.380M in   5.013447s
        fast_version      3.988M (±10.1%) i/s -     19.847M in   5.039580s

Comparison:
        fast_version:  3988340.3 i/s
      master_version:  2683336.2 i/s - 1.49x  slower

delete
master_version
{:FREE=>-5003, :T_STRING=>3003, :T_MATCH=>999, :T_IMEMO=>1000}
fast_version
{:FREE=>-3002, :T_STRING=>1001, :T_MATCH=>1000, :T_IMEMO=>1000}
Warming up --------------------------------------
      master_version    47.221k i/100ms
        fast_version    44.153k i/100ms
Calculating -------------------------------------
      master_version    597.881k (±11.4%) i/s -      2.975M in   5.047200s
        fast_version    686.014k (±11.6%) i/s -      3.400M in   5.036564s

Comparison:
        fast_version:   686014.5 i/s
      master_version:   597881.4 i/s - same-ish: difference falls within error
```

oniofchaos commented:
Squashed!

eugeneius commented:
Did you investigate @matthewd's suggestion here? https://github.com/rails/rails/pull/30963#discussion_r146381317

It seems like a simpler way to save the same number of allocations.

oniofchaos commented:
Took another look at that comment and tested it out and looks like it provides a 2x speedup compared to my current solution for non-get arguments such as `delete` but a 10x slowdown for `get`. Not sure if that's worth? Or maybe there's a further optimization to make based on using `match`?

Updated benchmarking code and results:
```ruby
# frozen_string_literal: true
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "benchmark-ips"
  gem "rails"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

@html_options = {}

def master_version
  if @method && @method.to_s.downcase != "get" && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

def fast_version
  if method_not_get_method?(@method) && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

def match_version
  if @method && /\Aget\z/i.match?(@method) && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

def match_version_with_const
  comp_method = (STRINGIFIED_COMMON_METHODS[@method] || @method)
  if @method && /\Aget\z/i.match?(comp_method) && @html_options["rel".freeze] !~ /nofollow/
    @html_options["rel".freeze] = "#{@html_options["rel".freeze]} nofollow".lstrip
  end
  @html_options["data-method".freeze] = @method
end

STRINGIFIED_COMMON_METHODS = {
  get:    'get',
  delete: 'delete',
  patch:  'patch',
  post:   'post',
  put:    'put',
}.freeze

def method_not_get_method?(method)
  return false unless method
  (STRINGIFIED_COMMON_METHODS[method] || method.to_s.downcase) != 'get'
end

def test
  puts "master_version"
  puts allocate_count { 1000.times { master_version } }
  puts "fast_version"
  puts allocate_count { 1000.times { fast_version } }
  puts "match_version"
  puts allocate_count { 1000.times { match_version } }
  puts "match_version_with_const"
  puts allocate_count { 1000.times { match_version_with_const } }

  Benchmark.ips do |x|
    x.report("master_version") { master_version }
    x.report("fast_version")     { fast_version }
    x.report("match_version")     { match_version }
    x.report("match_version_with_const")     { match_version_with_const }
    x.compare!
  end
end

puts 'get'
@method = :get

test

puts 'delete'
@method = :delete

test
```ruby
get
master_version
{:FREE=>-1932, :T_STRING=>2052}
fast_version
{:FREE=>-1}
match_version
{:FREE=>-5003, :T_STRING=>1003, :T_MATCH=>1999, :T_IMEMO=>2000}
match_version_with_const
{:FREE=>-5003, :T_STRING=>1002, :T_MATCH=>2000, :T_IMEMO=>2000}
Warming up --------------------------------------
      master_version   117.834k i/100ms
        fast_version   170.545k i/100ms
       match_version    33.021k i/100ms
match_version_with_const
                        35.210k i/100ms
Calculating -------------------------------------
      master_version      2.513M (± 9.4%) i/s -     12.490M in   5.017587s
        fast_version      4.038M (± 8.2%) i/s -     20.124M in   5.022930s
       match_version    385.497k (± 9.5%) i/s -      1.915M in   5.011259s
match_version_with_const
                        398.243k (±11.8%) i/s -      1.972M in   5.024023s

Comparison:
        fast_version:  4037895.6 i/s
      master_version:  2512706.7 i/s - 1.61x  slower
match_version_with_const:   398243.2 i/s - 10.14x  slower
       match_version:   385496.7 i/s - 10.47x  slower

delete
master_version
{:FREE=>-5002, :T_STRING=>3001, :T_MATCH=>1000, :T_IMEMO=>1000}
fast_version
{:FREE=>-3002, :T_STRING=>1001, :T_MATCH=>1000, :T_IMEMO=>1000}
match_version
{:FREE=>-1001, :T_IMEMO=>1000}
match_version_with_const
{:FREE=>-1001, :T_IMEMO=>1000}
Warming up --------------------------------------
      master_version    43.377k i/100ms
        fast_version    58.810k i/100ms
       match_version    98.709k i/100ms
match_version_with_const
                        96.942k i/100ms
Calculating -------------------------------------
      master_version    629.208k (± 9.9%) i/s -      3.123M in   5.009754s
        fast_version    715.365k (± 9.1%) i/s -      3.587M in   5.052967s
       match_version      1.466M (± 6.0%) i/s -      7.304M in   5.000822s
match_version_with_const
                          1.437M (± 7.5%) i/s -      7.174M in   5.025284s

Comparison:
       match_version:  1466121.0 i/s
match_version_with_const:  1436722.2 i/s - same-ish: difference falls within error
        fast_version:   715365.0 i/s - 2.05x  slower
      master_version:   629208.3 i/s - 2.33x  slower
```

rafaelfranca commented:
Usually the method is something different of `:get` so the `match` version seems to be better.

matthewd commented:
You're missing a `!` -- my proposal is slower in both. Looks like setting up the match is just too much work.
